2020-05-16-password-2.0.md

```diff
@@ -1,10 +1,246 @@
+[Bcrypt](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html).
+The trade-offs of the different algorithms are discussed in more detail
+[in the Haddocks](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password.html).
+
+Once you've picked the hashing algorithm, you can import its module.  We use
+[`Data.Password.Bcrypt`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html)
+as an example:
+
+```haskell
+> import Data.Password.Bcrypt
+```
+
+Next, we need a
+[`Password`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html#t:Password)
+to work with.  Let's create one with
+[`mkPassword`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html#v:mkPassword)[^2]:
+
+```haskell
+> let pass = mkPassword "foobar"
+> pass :: Password
+**PASSWORD**
+```
+
+Note that the `Show` instance for `Password` doesn't actually show the
+password.  This is a security measure.  You don't have to worry about passwords
+leaking into your web application logs.
+
+Now that you have a password, we can hash it with the
+[`hashPassword`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html#v:hashPassword)
+function.  This produces a
+[`PasswordHash`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html#t:PasswordHash):
+
+```haskell
+> hash <- hashPassword pass
+> hash :: PasswordHash Bcrypt
+PasswordHash {unPasswordHash = "$2b$10$IbnlktUyGjeunSlE4bQtfu89LY6veyDW0CGIoR5Kj8qrQa916txMS"}
+```
+
+This `PasswordHash` can be stored in a database[^3].
+
+The
+[`checkPassword`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password-Bcrypt.html#v:checkPassword)
+function can be used to check a `Password` against a `PasswordHash`:
+
+```haskell
+> let result = checkPassword pass hash
+> result :: PasswordCheck
+PasswordCheckSuccess
+```
+
+## Future Work
+
+There are a few issues in the `password` library that we'd like help with from
+anyone interested!
+
+-   Split off the
+    [`Data.Password`](https://hackage.haskell.org/package/password-2.0.1.1/docs/Data-Password.html)
+    module into a separate package called `password-types`.
+
+    All of the hashing algorithms provided in `password` come from
+    [`cryptonite`](https://hackage.haskell.org/package/cryptonite).
+
+    However, we would ideally like people to be able to use the `Password` and
+    `PasswordHash` data types as canonical data types in different libraries on
+    Hackage.  We don't want all these packages to have to pull in a transitive
+    dependency on `cryptonite`.
+
+    Ideally, we would have a `password-types` package that just exports
+    everything from the current `Data.Password` module (which doesn't require a
+    dependency on `cryptonite`).
+
+    https://github.com/cdepillabout/password/issues/20
+
+-   Add more password hashing algorithms.
```

Comment:
I'm not really sure we NEED more password hashing algorithms.
The 4 major ones are good enough IMHO (and PBKDF2 is only in there for legacy reasons, I'd never advise someone to use it for a new application), though maybe more algorithms could be added for that same reason, just legacy/interoperability... 🤔 

I WOULD like to add to this list the possibility to parse different formats of hashes for the same algorithm, so that `checkPassword` works with hashes produced by other (language's) libraries/programs.
(Maybe give a link to [this hashcat page](https://hashcat.net/wiki/doku.php?id=example_hashes) to show examples)

I'm not yet sure if it is a good idea to also have the format be adjustable in the `hashPassword` function... I don't see that having a good use case... but if someone would have a good use case, that'd be something that also COULD be worked on.

Reply:
Sounds good.  I removed the item for adding more password hashing algorithms, and I added a new item for parsing different formats.

I didn't link to the hashcat page, but I did link to https://github.com/cdepillabout/password/issues/11.  (If people click through, they will find the link to the hashcat page as well.)